### PR TITLE
Add MPEG-H support

### LINKED ISSUE #930: Support for MHA1 audio on packager
## System info
**Operating System**: <e.g. macOS Sierra, Ubuntu 14.04 trusty etc>
`AmazonLinux`
**Shaka Packager Version**: <e.g. v1.6.1, commit SHA etc>
[6c8ad30217c286d4eecadc9df12420767d389942](https://github.com/google/shaka-packager/commit/6c8ad30217c286d4eecadc9df12420767d389942)

## Issue and steps to reproduce the problem

**Packager Command**:
```
docker run -it -v /tmp:/media local_registry:DigitalMusicShakaPackager-1.0-1618201405.1419096 packager "in=/media/TempMHA1.mp4,stream=audio,output=/media/output/TempMHA1.mp4" --fragment_duration 10 --segment_duration 10 --clear_lead 10 --v=-1 --generate_static_mpd --mpd_output /media/output/manifest-3.mpd
```

Extra steps to reproduce the problem?
(1) Run with MP4 content that contains MHA1 audio.

What is the expected result?
Would like support for the MHA1 audio type

What happens instead?
Error with unsupported audio format, see error log:

```
[0412/222408:WARNING:mp4_media_parser.cc(519)] Unsupported audio format 'mha1' in stsd box.
[0412/222408:ERROR:audio_stream_info.cc(143)] Not implemented reached in static std::string shaka::media::AudioStreamInfo::GetCodecString(shaka::media::Codec, uint8_t)Codec: 0
[0412/222408:ERROR:mp4_muxer.cc(517)] Not implemented reached in bool shaka::media::mp4::MP4Muxer::GenerateAudioTrak(const shaka::media::AudioStreamInfo *, shaka::media::mp4::Track *) Unsupported audio codec 0
[0412/222408:ERROR:demuxer.cc(356)] Failed to process sample 512 11 (MUXER_FAILURE): Failed to generate trak.
[0412/222408:ERROR:mp4_media_parser.cc(879)] Failed to process the sample.
[0412/222408:ERROR:packager_main.cc(550)] Packaging Error: 8 (PARSER_FAILURE): Cannot parse media file /media/TempMHA1.mp4
```

<Please attach the input files or email to shaka-packager-issues@google.com.>
<Please provide any additional information below.>

### PR DESCRIPTION
**Pull request by nvincen**
Continuation from #933, and closes #930.

I messed up my pull request.

### CODE REVIEW

File: packager/media/formats/mp4/box_definitions.h [hunk 1]
```diff
@@ -382,6 +389,13 @@ struct AudioSampleEntry : Box {
   AC4Specific dac4;
   OpusSpecific dops;
   FlacSpecific dfla;
+  MHAConfiguration mhac;
+
+  //23008-3 decoder specific config
+  CodecConfiguration codec_configuration;
+
+  // Returns the box type of codec configuration box from audio format.
+  FourCC GetCodecConfigurationBoxType(FourCC format) const;
```
Comment (nvincen) [inline packager/media/formats/mp4/box_definitions.h:398]: MPEG-H decoders need a codec_configuration values, similar to video codecs. So I added a codec_configuration box to allow for the holding and parsing of the configuration_stream.

If there is a better way to implement, please suggest.

File: packager/media/formats/mp4/box_definitions.cc [hunk 2]
```diff
@@ -1709,6 +1709,29 @@ size_t ElementaryStreamDescriptor::ComputeSizeInternal() {
...
+FourCC MHAConfiguration::BoxType() const {
+  return FOURCC_mhaC;
+}
+
+bool MHAConfiguration::ReadWriteInternal(BoxBuffer* buffer) {
+  RCHECK(ReadWriteHeaderInternal(buffer) &&
+         buffer->ReadWriteVector(
+             &data, buffer->Reading() ? buffer->BytesLeft() : data.size()));
+  RCHECK(data.size() > 1);
+  mpeg_h_3da_profile_level_indication = data[1];
```
Comment (nvincen) [inline packager/media/formats/mp4/box_definitions.cc:1724]: Should this be inside an `if (buffer->Reading())` condition? i.e.

```
  if (buffer->Reading()) {
    RCHECK(data.size() > 1);
    mpeg_h_3da_profile_level_indication = data[1];
  }
```

**Comment by nvincen** [comment]
@kqyang - Would you mind having another look?

**Comment by nvincen** [comment]
@kqyang - Friendly ping to review the latest changes

File: packager/media/base/audio_stream_info.cc [hunk 3]
```diff
@@ -139,6 +139,18 @@ std::string AudioStreamInfo::GetCodecString(Codec codec,
...
+      // The signalling of the codecs parameters is according to RFC6381 [11] and ISO/IEC 23008-3 clause 21 [7].
+      // The value consists of the following two parts separated by a dot:
+      //  - the sample entry 4CC code ('mha1', 'mha2', 'mhm1', 'mhm2')
+      //  - ‘0x’ followed by the hex value of the profile-levelid, as defined in in ISO/IEC 23008-3 [7]
+      return base::StringPrintf("mha1.0x%02x", audio_object_type);
+    case kCodecMhm1:
+      // The signalling of the codecs parameters is according to RFC6381 [11] and ISO/IEC 23008-3 clause 21 [7].
+      // The value consists of the following two parts separated by a dot:
+      //  - the sample entry 4CC code ('mha1', 'mha2', 'mhm1', 'mhm2')
+      //  - ‘0x’ followed by the hex value of the profile-levelid, as defined in in ISO/IEC 23008-3 [7]
```
Comment (kqyang) [inline packager/media/base/audio_stream_info.cc:152]: I'd remove the duplicated comments and replace it with:

```
// See above.
```

Or even better, combine the two cases:

```
case kCodecMha1:
case kCodecMhm1:
  ...
  return base::StringPrintf("%s.0x%02x", FourCCToString (CodecToFourCC(codec)), audio_object_type);
```
```
FourCC CodecToFourCC(Codec codec) {
  switch (codec) {
    case kCodecMha1:
      return FOURCC_mha1;
    case kCodecMhm1:
      return FOURCC_mhm1;
    default:
      return FOURCC_NULL;
  }
```
}

File: packager/media/base/stream_info.h [hunk 4]
```diff
@@ -55,6 +55,8 @@ enum Codec {
   kCodecVorbis,
   kCodecMP3,
   kCodecAudioMaxPlusOne,
+  kCodecMha1,
+  kCodecMhm1,
```
Comment (kqyang) [inline packager/media/base/stream_info.h:59]: Can you move them before kCodecAudioMaxPlusOne, i.e. before line 57?

File: packager/media/formats/mp4/box_definitions.cc (hunk 2 above)
Comment (kqyang) [reply packager/media/formats/mp4/box_definitions.cc:1724]: Yes, that is better I think, though it does not really matter.

File: packager/media/formats/mp4/box_definitions.h [hunk 5]
```diff
@@ -328,6 +328,13 @@ struct AC3Specific : Box {
   std::vector<uint8_t> data;
 };
 
+struct MHAConfiguration : Box {
+  DECLARE_BOX_METHODS(MHAConfiguration);
+
+  std::vector<uint8_t> data;
+  std::uint8_t mpeg_h_3da_profile_level_indication;
```
Comment (kqyang) [inline packager/media/formats/mp4/box_definitions.h:335]: Replace std::uint8_t with uint8_t to be consistent

File: packager/media/formats/mp4/box_definitions.h (hunk 1 above)
Comment (kqyang) [reply packager/media/formats/mp4/box_definitions.h:398]: Shouldn't it be handled in `MHAConfiguration mhac` already? Isn't the mhac the codec configuration for MPEG-H? 

I understand that it is a bit confusing, but the approach for audio and video are slightly different. For audio, we are listing all the possible audio decoder configurations while for video, we are using the same CodecConfiguration structure. The code should really be cleaned up.

I think we can just remove line 393 to 398. The functionality has already been implemented in `mhac`. Let me know if I misunderstood your intention and sorry for the confusion.

**Review by kqyang — COMMENTED** [review]
Thanks Vincent. Sorry for the delay in the code review. The CL looks mostly good to me with a few more comments (with most of them as style nits).

File: packager/media/formats/mp4/box_definitions.h (hunk 5 above)
Comment (nvincen) [reply packager/media/formats/mp4/box_definitions.h]: Done

File: packager/media/formats/mp4/box_definitions.h (hunk 1 above)
Comment (nvincen) [reply packager/media/formats/mp4/box_definitions.h]: Removed 393 - 398

File: packager/media/formats/mp4/box_definitions.cc (hunk 2 above)
Comment (nvincen) [reply packager/media/formats/mp4/box_definitions.cc:1724]: Ok, I just left it alone.

File: packager/media/base/audio_stream_info.cc [hunk 6]
```diff
@@ -139,6 +139,18 @@ std::string AudioStreamInfo::GetCodecString(Codec codec,
       return "mp3";
     case kCodecVorbis:
       return "vorbis";
+    case kCodecMha1:
+      // The signalling of the codecs parameters is according to RFC6381 [11] and ISO/IEC 23008-3 clause 21 [7].
+      // The value consists of the following two parts separated by a dot:
+      //  - the sample entry 4CC code ('mha1', 'mha2', 'mhm1', 'mhm2')
+      //  - ‘0x’ followed by the hex value of the profile-levelid, as defined in in ISO/IEC 23008-3 [7]
```
[the remark this answers is not in the document]
Comment (nvincen) [reply packager/media/base/audio_stream_info.cc]: Addressed with below suggestion.

File: packager/media/base/audio_stream_info.cc (hunk 3 above)
Comment (nvincen) [reply packager/media/base/audio_stream_info.cc]: Merged the two cases.

File: packager/media/base/stream_info.h (hunk 4 above)
Comment (nvincen) [reply packager/media/base/stream_info.h:58]: Done.

**Review by kqyang — COMMENTED** [review]
Thanks. Two more style comments.

File: packager/media/base/audio_stream_info.cc [hunk 7]
```diff
@@ -140,17 +151,14 @@ std::string AudioStreamInfo::GetCodecString(Codec codec,
...
-      // The signalling of the codecs parameters is according to RFC6381 [11] and ISO/IEC 23008-3 clause 21 [7].
+      // The signalling of the codecs parameters is according to RFC6381 [11]
+      //   and ISO/IEC 23008-3 clause 21 [7].
       // The value consists of the following two parts separated by a dot:
       //  - the sample entry 4CC code ('mha1', 'mha2', 'mhm1', 'mhm2')
-      //  - ‘0x’ followed by the hex value of the profile-levelid, as defined in in ISO/IEC 23008-3 [7]
-      return base::StringPrintf("mhm1.0x%02x", audio_object_type);
+      //  - ‘0x’ followed by the hex value of the profile-levelid, as defined 
+      //      in in ISO/IEC 23008-3 [7]
+      return base::StringPrintf("%s.0x%02x", FourCCToString(CodecToFourCC(codec)).c_str(), audio_object_type);
```
[the remark this answers is not in the document]
Comment (nvincen) [reply packager/media/base/audio_stream_info.cc]: Apologies, I missed that one. Fixed.

File: packager/media/formats/mp4/box_definitions_unittest.cc [hunk 8]
```diff
@@ -1222,6 +1222,36 @@ TEST_F(BoxDefinitionsTest, AC3SampleEntry) {
...
+  entry.samplesize = 16;
+  entry.samplerate = 44100;
+  Fill(&entry.mhac);
+  entry.Write(this->buffer_.get());
+
+  AudioSampleEntry entry_readback;
+  ASSERT_TRUE(ReadBack(&entry_readback));
+  ASSERT_EQ(entry, entry_readback);
+}
+
```
[the remark this answers is not in the document]
Comment (nvincen) [reply packager/media/formats/mp4/box_definitions_unittest.cc]: Removed.

**Review by kqyang — COMMENTED** [review]
One more question: can you generate a small test content (~3 seconds) and create an end to end test, similar to https://github.com/google/shaka-packager/blob/master/packager/app/test/packager_test.py#L1022?

You can then run:

```
out/Release/packager_test.py PackagerFunctionalTest.testMpegHMp4 --test_update_golden_files
```
to update the golden files.

This step is optional but good to have. It is fine to skip it if you cannot fine or generate a copy right free content easily.

File: packager/media/base/fourccs.h [hunk 9]
```diff
@@ -87,6 +87,9 @@ enum FourCC : uint32_t {
   FOURCC_meco = 0x6d65636f,
   FOURCC_mehd = 0x6d656864,
   FOURCC_meta = 0x6d657461,
+  FOURCC_mha1 = 0x6d686131,
+  FOURCC_mhaC = 0x6d686143,
+  FOURCC_mhm1 = 0x6d686d31,
```
[the remark this answers is not in the document]
Comment (nvincen) [reply packager/media/base/fourccs.h:94]: Arranged to alphabetical order.

**Comment by nvincen** [comment]
> One more question: can you generate a small test content (~3 seconds) and create an end to end test, similar to https://github.com/google/shaka-packager/blob/master/packager/app/test/packager_test.py#L1022?
>
> You can then run:
> ```
> out/Release/packager_test.py PackagerFunctionalTest.testMpegHMp4 --test_update_golden_files 
> ```
> to update the golden files.
> 
> This step is optional but good to have. It is fine to skip it if you cannot fine or generate a copy right free content easily.

I agree it's good to have a small file added. I unfortunately don't have any copyright-free content handy. I'm going to get in touch with Sony to get my hands on some test content that I can upload. Can that be uploaded in a future PR?

**Comment by kqyang** [comment]
Thanks for adding the end to end test. Can you generate a smaller file, i.e. take the first three seconds only?

**Comment by nvincen** [comment]
> Thanks for adding the end to end test. Can you generate a smaller file, i.e. take the first three seconds only?

This was a file that I found floating around on the web. Do you know how I could trim the file to 3 seconds? I wasn't able to get ffmpeg to trim the file.

**Comment by kqyang** [comment]
> This was a file that I found floating around on the web. 

Does it explicitly indicate that it can be shared freely? If not, we won't be able to use it in Shaka Packager.

> Do you know how I could trim the file to 3 seconds? I wasn't able to get ffmpeg to trim the file.

Does the below command work?

```
ffmpeg -ss 00:00:00 -i input.mp4 -to 00:00:03 -c copy output.mp4
```

If not, it's fine not to having end to end test.

**Comment by nvincen** [comment]
> > This was a file that I found floating around on the web.
> 
> Does it explicitly indicate that it can be shared freely? If not, we won't be able to use it in Shaka Packager.
> 
> > Do you know how I could trim the file to 3 seconds? I wasn't able to get ffmpeg to trim the file.
> 
> Does the below command work?
> 
> ```
> ffmpeg -ss 00:00:00 -i input.mp4 -to 00:00:03 -c copy output.mp4
> ```
> 
> If not, it's fine not to having end to end test.

It looks like it's copyrighted actually:

```
bash-4.2# ffmpeg -i /media/tmp/mha1.mp4
ffmpeg version 4.3.1 Copyright (c) 2000-2020 the FFmpeg developers
  built with gcc 7 (GCC)
  configuration: --prefix=/local/p4clients/pkgbuild-U9Jv8/workspace/src/FFmpeg/build/private/install --disable-static --enable-shared --enable-encoder=vorbis --enable-libvorbis --enable-libmp3lame --enable-libvpx --enable-libaom --enable-libfdk-aac --enable-libopus --enable-openssl
  libavutil      56. 51.100 / 56. 51.100
  libavcodec     58. 91.100 / 58. 91.100
  libavformat    58. 45.100 / 58. 45.100
  libavdevice    58. 10.100 / 58. 10.100
  libavfilter     7. 85.100 /  7. 85.100
  libswscale      5.  7.100 /  5.  7.100
  libswresample   3.  7.100 /  3.  7.100
[mov,mp4,m4a,3gp,3g2,mj2 @ 0x5555557c2c40] stream 0, timescale not set
[mov,mp4,m4a,3gp,3g2,mj2 @ 0x5555557c2c40] Could not find codec parameters for stream 0 (Audio: none (mha1 / 0x3161686D), 48000 Hz, 0 channels, 666 kb/s): unknown codec
Consider increasing the value for the 'analyzeduration' and 'probesize' options
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from '/media/tmp/mha1.mp4':
  Metadata:
    major_brand     : mp42
    minor_version   : 0
    compatible_brands: mp42isom
    creation_time   : 2020-08-05T18:08:52.000000Z
    album_artist    : Pharrell Williams
    track           : 1
    date            : 2020
    disc            : 1/1
    copyright       : (P) 2013 Back Lot Music, under exclusive license to Columbia Records, a Division of Sony Music Entertainment
    artist          : Pharrell Williams
    title           : Happy
    composer        : Pharrell Williams
    album           : Pharrell Williams 360 Reality Audio [360 Reality Audio MPEG-H Tidal]
  Duration: 00:03:54.36, start: 0.000000, bitrate: 680 kb/s
    Stream #0:0(und): Audio: none (mha1 / 0x3161686D), 48000 Hz, 0 channels, 666 kb/s (default)
    Metadata:
      creation_time   : 2020-08-05T18:08:52.000000Z
      handler_name    : soun
    Stream #0:1: Video: mjpeg (Progressive), yuvj444p(pc, bt470bg/unknown/unknown), 1280x1280 [SAR 300:300 DAR 1:1], 90k tbr, 90k tbn, 90k tbc (attached pic)
At least one output file must be specified
```

My last commit reverts the E2E test.

**Comment by kqyang** [comment]
@nvincen I have just merged the change. Thanks for working on this feature and supporting Shaka Packager project!

**Comment by ronak2121** [comment]
@kqyang is there any ETA on when this support will be available in a release?

**Comment by ronak2121** [comment]
Hi @joeyparrish any update on when this support would be in a release?

**Comment by joeyparrish** [comment]
I need to check that our new low-latency feature is ready to ship.  Assuming it is, I could release v2.6.0 within the next week.  Monday is a holiday in the US, so possibly Tuesday if LL is ready.

**Comment by joeyparrish** [comment]
Looks like it's ready.  I'm going to start the release process.

**Comment by joeyparrish** [comment]
https://github.com/google/shaka-packager/releases/tag/v2.6.0

**Comment by ronak2121** [comment]
Awesome thank you!

**Pull request merged**